### PR TITLE
Python/Local Hugging Face Inference for Completions and Embeddings

### PR DESCRIPTION
**Pull request by awharrison-28**
### Motivation and Context
This PR introduces native python support for Hugging Face models that can: complete text, generate new text, summarize, and that generate embeddings. Currently only supports downloading the models locally from the HF model hub. Future plans include supporting the HF inference API as well.

### Description
- Added 2 services: `hf_text_completion` and `hf_text_embedding`
- `hf_text_completion` supports the following tasks: _text-generation_, _text2text-generation_, and _summarization_
- `hf_text_embedding` supports any model supported by the sentence-transformers pip package
- Added dependencies: pytorch, transformers, sentence-transformers to `requirements.txt` and `poetry.lock`
- fixed typo: `get_embedding_service_service_id` -> `get_embedding_service_id`
- Added a number of integration tests for supported HF models 

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: FEATURE_MATRIX.md [hunk 1]
```diff
@@ -13,8 +13,9 @@
 |---|---|---|---|
 | OpenAI                            | ✅ | ✅ | |
 | AzureOpenAI                       | ✅ | ✅ | |
-| Hugging Face                      | ✅ | ❌ | Coming soon to Python - both native and web endpoint support |
-| Custom                            | ✅ | ❌ | Requires the user to define the service schema in their application |
+| Hugging Face Inference API        | 🔄 | ❌ | Coming soon to Python, not all scenarios are covered for .NET |
+| Hugging Face Local                | ❌ | ✅ | |
```
Comment (shawncal) [inline FEATURE_MATRIX.md:17]: Ohhh, not fair. Surely C# can at least get a "N/A" for "Hugging Face Local"

Comment (awharrison-28) [reply FEATURE_MATRIX.md:17]: We're aiming to effectively support via some combination of TorchSharp, ML.NET, and ONNXRuntime. I think this is fair :)

File: python/semantic_kernel/connectors/ai/ai_exception.py [hunk 2]
```diff
@@ -28,6 +28,8 @@ class ErrorCodes(Enum):
         InvalidConfiguration = 8
         # The function is not supported.
         FunctionTypeNotSupported = 9
+        # The HF task is not supported.
+        UnsupportedHFTask = 10
```
Comment (dluc) [inline python/semantic_kernel/connectors/ai/ai_exception.py:32]: could we avoid branded error codes, maybe reuse an existing code or introduce one that can be reused? you can use error messages to provide further details

File: python/semantic_kernel/connectors/ai/hugging_face/services/hf_text_completion.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        self._task = "text2text-generation" if task is None else task
+        self._log = log if log is not None else NullLogger()
+        self.device = (
+            "cuda:" + device if device >= 0 and torch.cuda.is_available() else "cpu"
+        )
+        self.generator = pipeline(
+            task=self._task, model=self._model_id, device=self.device
+        )
+
+    async def complete_simple_async(
```
Comment (dluc) [inline python/semantic_kernel/connectors/ai/hugging_face/services/hf_text_completion.py:58]: what is "simple" about this method, vs... something else maybe ?

Comment (awharrison-28) [reply python/semantic_kernel/connectors/ai/hugging_face/services/hf_text_completion.py:58]: I called it simple because the semantic function is only {{$input}} - I'll give it a better name though

Comment (awharrison-28) [reply python/semantic_kernel/connectors/ai/hugging_face/services/hf_text_completion.py:58]: I think it was called simple because it doesn't batch? I can rename it in the base class to something better

File: python/semantic_kernel/connectors/ai/ai_exception.py (hunk 2 above)
Comment (awharrison-28) [reply python/semantic_kernel/connectors/ai/ai_exception.py:32]: reusing ServiceError instead

File: python/semantic_kernel/connectors/ai/hugging_face/services/hf_text_completion.py (hunk 3 above)
Comment (awharrison-28) [reply python/semantic_kernel/connectors/ai/hugging_face/services/hf_text_completion.py:58]: renamed to complete_async to match .NET. Behavior is in sync, so no reason for the methods to be named differently

File: FEATURE_MATRIX.md [hunk 4]
```diff
@@ -13,13 +13,14 @@
...
+| Custom                            | ✅ | 🔄 | Requires the user to define the service schema in their application |
 
 ## Tokenizers
 | | C# | Python | Notes |
 |---|---|---|---|
-| GPT2                              | ✅ | 🔄 | Can be manually added to Python via `pip install transformers` |
+| GPT2                              | ✅ | ✅ | |
```
Comment (lemillermicrosoft) [inline FEATURE_MATRIX.md:23]: Just curious, I'm guessing this is unrelated to this change, what PR enabled this?

File: python/requirements.txt [hunk 5]
```diff
@@ -1,3 +1,6 @@
 openai==0.27.*
 numpy==1.24.*
-aiofiles>=23.1.0
\ No newline at end of file
+aiofiles>=23.1.0
```
Comment (lemillermicrosoft) [inline python/requirements.txt:3]: I'm surprised to not see these locked down to at least a major version. I've had problems with this in the past but maybe I'm worrying too much. Thoughts?

**Review by lemillermicrosoft — APPROVED** [review]
Few comments that you may resolve as appropriate.

File: python/semantic_kernel/connectors/ai/ai_exception.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lemillermicrosoft) [reply python/semantic_kernel/connectors/ai/ai_exception.py]: Oops I was late to publishing, Devis called this out as well.

File: FEATURE_MATRIX.md (hunk 4 above)
Comment (awharrison-28) [reply FEATURE_MATRIX.md:23]: This is related. This PR introduces `transformers` as a formal dependency of Python SK. You can register a HuggingFaceTextCompletion('gpt2') connection to the kernel now.

File: python/requirements.txt (hunk 5 above)
Comment (awharrison-28) [reply python/requirements.txt:3]: currently is not an issue for our CI, taking down a task to better pin our dependencies

File: python/semantic_kernel/connectors/ai/hugging_face/services/hf_text_completion.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+            if self._task == "text-generation" or self._task == "text2text-generation":
+                return result[0]["generated_text"]
+
+            elif self._task == "summarization":
+                return result[0]["summary_text"]
+
+            else:
+                raise AIException(
+                    AIException.ErrorCodes.ServiceError,
```
Comment (dluc) [inline python/semantic_kernel/connectors/ai/hugging_face/services/hf_text_completion.py:89]: wondering, is "ServiceError" the right code? looks like ServiceError would be an external error, but here it's more like a user (developer) error

Comment (awharrison-28) [reply python/semantic_kernel/connectors/ai/hugging_face/services/hf_text_completion.py:89]: I was a torn between ServiceError, ModelNotAvailable, and InvalidConfiguration. 

I leaned towards ServiceError because it's the service that label's the model as being capable of a 'pipeline task'

Thoughts? Maybe InvalidConfiguration makes more sense since the user is passing the 'pipeline task' as completion configuration?

Comment (awharrison-28) [reply python/semantic_kernel/connectors/ai/hugging_face/services/hf_text_completion.py:89]: changed to InvalidConfiguration and made the error message more verbose about the supported pipeline tasks

**Review by dluc — APPROVED** [review]
